Basic Docker+Vagrant setup for dev of MTP

mikepea opened the pull request:
This adds an experimental Docker setup within a Vagrant virtualbox
using the vagrant-cachier and -- most experimentally -- vagrant-persistent-storage
plugins to speed up the vagrant destroy/up process.

vagrant-persistent-storage is configured to use a local 5GB virtualbox disk image to
store the /var/lib/docker dir (within the VM) across redeploys.

This prevents the need for redownload of images/layers from the Docker Hub.

mikepea committed:
Basic Docker+Vagrant setup for dev of MTP

This adds an experimental Docker setup within a Vagrant virtualbox
using the vagrant-cachier and -- most experimentally -- vagrant-persistent-storage
plugins to speed up the vagrant destroy/up process.

vagrant-persistent-storage is configured to use a local 5GB virtualbox disk image to
store the /var/lib/docker dir (within the VM) across redeploys.

This prevents the need for redownload of images/layers from the Docker Hub.

ashb commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+RUN touch /etc/inittab
+
+# runit setup for the application
+RUN mkdir -p /etc/service/prison-visits
```
This is not prison-visits anymore Toto ;)

mikepea replied:
I :heart: :apple:-V

SteveMarshall replied:
I don't see an :apple: on my keyboard, only a ⌘? :wink:

SteveMarshall commented:
@milosgajdos83 @ashb Can you two cast an eye over this with me? You're probably a bit more closely acquainted with this… Ta!

SteveMarshall commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+unless Vagrant.has_plugin?("vagrant-cachier")
+  puts "WARNING: vagrant-cachier plugin is not installed! It really speeds this up..."
+  puts "         Install using 'vagrant plugin install vagrant-cachier'"
+end
```
Can we log these in such a way that they look even more like a warning? Maybe colourise them, or use whatever Vagrant uses for warnings? ([Looks like that is `Log4r::Logger`](https://github.com/mitchellh/vagrant/blob/efd1d5e11bfc5a72c7a1d1eae294b4751d841544/plugins/synced_folders/nfs/action_cleanup.rb#L8)?)

mikepea replied:
I'm not going to worry about this for now - had a quick go with Log4r::Logger and started to go down a rabbit hole. There's something unusual with how vagrant evaluates the Vagrantfile.

SteveMarshall replied:
:+1:

SteveMarshall commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+ENV['VAGRANT_DEFAULT_PROVIDER'] ||= "virtualbox"
```
:+1:

ashb replied:
:-1:  I like VMware cos it's faster. Why have this and not let Vagrant use what ever it determines is best?

SteveMarshall replied:
Wait… It can do that now? I thought it defaulted to Virtualbox always? Oooooh. In that case, wot @ashb sez.

SteveMarshall commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+end
+
+unless Vagrant.has_plugin?("vagrant-persistent-storage")
+  puts "WARNING: vagrant-persistent-storage not installed! It really speeds up docker image mgmt"
```
_koff_ 80-character lines _koff_?

Also: [mgmt](https://en.wikipedia.org/wiki/MGMT)? :stuck_out_tongue_closed_eyes:

mikepea replied:
We have an 80-char limit in our coding standards? Care to point me to them? :p

SteveMarshall replied:
We may not, but `mgmt` made it look like you were adhering to one (or something similar)… But also, 80 character lines are nice, m'kay? (I'm not actually saying we should adhere religiously to one, but if a nice can be nicely made to fit, why not?)

mikepea committed:
Add a runit script ready for when we have an app

Handle the case where the Gemfile is not present, so that
don't throw an error.

mikepea committed:
Increase storage for vagrant-persistent-storage

5GB is not enough for a workable Docker image cache

ashb commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+  # Set up SSH agent forwarding.
+  config.ssh.forward_agent = true
+
+  if Vagrant.has_plugin?("vagrant-persistent-storage")
```
Hmmm this plugin _only_ works for Vbox. I guess the question is what does it do on vmware (does it fail to run or just have no effect)?

ashb commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+
+Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+  config.vm.box = "ubuntu/trusty64"
+  config.vm.hostname = "prison-visits"
```
Find-and-replace on "prison-visits" in this file too

ashb commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+  # print out help
+  config.vm.provision "shell", inline: <<-EOF
+    echo "---------------------------------------"
+    echo "export DOCKER_HOST=tcp://localhost:2376"
```
Use DOCKER_PORT rather than hard-code the port to 2376 please

ashb commented:
@mikepea What does vagrant-persistent-storage plugin give us over the vagrant-cachiers' [Generic buckets](http://fgrehm.viewdocs.io/vagrant-cachier/buckets/generic):

``` ruby
Vagrant.configure("2") do |config|
  config.cache.enable :generic, { :cache_dir => "/var/lib/docker }
end
```

ashb commented:
Also more generally: I worry that this caching might expose us to the docker devicemapper race condition more - I think Peter said that having /var/lib/docker as a mount caused the problem more.

ashb commented on Vagrantfile:
```diff
@@ -0,0 +1,62 @@
+end
+
+Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+  config.vm.box = "ubuntu/trusty64"
```
This image only supports virtualbox - Can we use https://vagrantcloud.com/puppetlabs/boxes/ubuntu-14.04-64-nocm instead as it supports both

SteveMarshall replied:
Alternatively, https://atlas.hashicorp.com/chef/boxes/ubuntu-14.04 for similar reasons.

SteveMarshall commented:
Just tried running `vagrant up` again, and am still getting this error:

```
# Executing 5 build triggers
Trigger 0, COPY Gemfile /usr/src/app/
Step 0 : COPY Gemfile /usr/src/app/
Trigger 1, COPY Gemfile.lock /usr/src/app/
Step 0 : COPY Gemfile.lock /usr/src/app/


Stderr from the command:

stdin: is not a tty
time="2015-01-23T09:57:28Z" level="info" msg="Gemfile.lock: no such file or directory" 
```

(I'm sure you're already working on this, but I figure noting it here is better than not.)

ashb commented:
@SteveMarshall That error is coming from the docker build process of our ruby-webapp base Dockerfile, and because this is _so_ MVP there isn't even an actually ruby app in here yet and it barfs ;)

SteveMarshall commented:
@ashb Well, maybe you and @milosgajdos83 should fix that? :wink:

ashb commented:
Sounds like an ecuminical matter.

More seriously: by having `FROM ministryofjustice/ruby:2.1.5-webapp-onbuild` you've said this is a ruby app, so there should be a Gemfile and a Gemfile.lock

SteveMarshall commented:
A fair point, I suppose. Just seems like it might be nice if it gave a better warning than blowing up without a lock file?

ashb commented:
It would be nice, but it's not possible. Yay docker!

SteveMarshall commented:
Oh, OK then. Yay docker. :disappointed:

mikepea committed:
Add a 'rails new .' skeleton

This is from rails 4.2, and should be pure.

mikepea committed:
Add vagrant to .gitignore

mikepea committed:
Remove vagrant-persistent-storage

As it was of limited value (only virtualbox) and is of questionable value in dev workflow

A local registry may be a better idea

mikepea committed:
Bind application and docker to 'unique' port

If we bind docker and the app to standard ports, then devs will
not be able to work on applications concurrently. Specifically,
the same dev team will be working on this project and prison-visits.

We'll need to keep track of these mappings somehow :/

mikepea committed:
Improve the help output from Vagrant

... to give details of why you'd set DOCKER_HOST, and
also how to access the running application

mikepea commented:
This now brings up a working Rails 4.2 skeleton, accessible via http://localhost:3001 so as not to conflict with the prison-visits Docker deploy (on the standard 3000 port)

@milosgajdos83 something i'm unsure about with this mechanism is how a dev would run 'bundle install' once the container is operational?

milosgajdos commented:
@mikepea you mean once the container is up and running ?

You can run `bundle install` via `docker exec -it`.

mikepea commented:
Will that work whilst rails is running? As-in, does Rails pick up the
updated libs in developer mode?

On 29 January 2015 at 10:48, Milos Gajdos notifications@github.com wrote:

> @mikepea https://github.com/mikepea you mean once the container is up
> and running ?
> 
> You can run bundle install via docker exec -it.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ministryofjustice/money-to-prisoners/pull/1#issuecomment-72005143
> .

ashb commented:
No. you'd need to restart the rails process (`docker stop mtp && docker start mtp`)  for that to get picked up, but installing new gems is a fairly rare occurrence.

Running `vagrant provision` might also make it pick up the new gems but could take longer as it would have to install all the gems again from fresh rather than just the new ones.

SteveMarshall commented:
@mikepea @ashb @milosgajdos83 Is there anything left to do on this before it's ready to merge?

ashb commented:
No nothing code wise - we need to try it and shake out the shards.

SteveMarshall commented:
Sooooo… can we :ship: this? Or wait till more platform stuff is done?

ashb commented:
@SteveMarshall :boom:
